### ISSUE TITLE
Ingestion fails on a transaction with a single EndSponsoringFutureReserves operation

### ISSUE THREAD

**Issue by bartekn** [stellar/go#2987]
### What version are you using?

`protocol-14` branch.

### What did you do?

Submit a single EndSponsoringFutureReserves (it will fail but will be included in a ledger):
```go
	tx, err := txnbuild.NewTransaction(
		txnbuild.TransactionParams{
			SourceAccount: source,
			Operations: []txnbuild.Operation{
				&txnbuild.EndSponsoringFutureReserves{
					SourceAccount: source,
				},
			},
			BaseFee:    txnbuild.MinBaseFee,
			Timebounds: txnbuild.NewInfiniteTimeout(),
		},
	)
```

### What did you expect to see?

No errors, operation ingested.

### What did you see instead?

```
Error running processors on ledger: Error streaming changes from ledger: could not process transaction 1: error in *processors.OperationProcessor.ProcessTransaction: Error obtaining details for operation 2353642082305: prior Begin operation not found for EndSponsoringFutureReserves
```

Remember that this makes `begin_sponsor` nullable and it needs to be changed in `protocols/horizon` and `services/horizon/resourceadapters`.

**Comment by 2opremio** [comment]
But ... isn't this a badly-formed transaction? What do we do in other situations like this?

**Comment by bartekn** [comment]
@2opremio it's a good point. I think it's due to the fact that Stellar-Core doesn't check validity of entire transaction, it's check per operation only. Anyway, if Stellar-Core accepts such transaction we need to ingest it. Unless, this changes in Stellar-Core: @jonjove want to check it out?

**Comment by jonjove** [comment]
@bartek is correct, stellar-core isn't validating transactions beyond the per-operation level. At some protocol meeting, we discussed whether stellar-core should reject bad pairings of begin/end sponsoring. The general consensus was that we didn't really need to do this, so we elected not to.

**Comment by abuiles** [comment]
@2opremio for effects we skip ingesting 

https://github.com/stellar/go/blob/f691bc858bd9b85d4958bba0154af149b3b87529/services/horizon/internal/expingest/processors/effects_processor.go#L110-L114

However, we don't do this for operation details. I think we'll need to add some special handling around the following lines

https://github.com/stellar/go/blob/f691bc858bd9b85d4958bba0154af149b3b87529/services/horizon/internal/expingest/processors/operations_processor.go#L343-L345

So it only fails if `beginSponsoringOp == nil` and the transaction was successful  -- otherwise we continue ingestion without adding any details which are based on other operations (like `BeingSponsoring..`)

We'll need to think about the different scenarios that can cause this and handle them.

**Comment by 2opremio** [comment]
I think we should still error. But use a special error which indicates it can be skipped if needed.

**Comment by abuiles** [comment]
@2opremio 

> I think we should still error. But use a special error which indicates it can be skipped if needed.

Why?  The behavior today is that if a tx fails -- we still show the operations in the envelope. 

If you send a `EndSponsoring` without a `Begin` you should still see it in the operations related with your TX but you'll only see `EndSponsoring` which is okay since the TX result was not successful.

**Comment by 2opremio** [comment]
> Why? The behavior today is that if a tx fails -- we still show the operations in the envelope.

Because otherwise we won't know there is a problem with the transaction. But maybe we don't care about that?

That's what I understood from your review comment at https://github.com/stellar/go/pull/2917#discussion_r473031893 , but maybe in hindsight you think we shouldn't error?

**Comment by bartekn** [comment]
> But maybe we don't care about that?

I think we don't care about this. Someone can very easily send a bunch of transactions like this in many ledgers and it will generate a bunch of errors.

**Comment by abuiles** [comment]
@2opremio 

> That's what I understood from your review comment at #2917 (comment) , but maybe in hindsight you think we shouldn't error?

I wrongly assumed stellar-core wouldn't include this tx in the ledger if it was invalid -- which is not the case. 

If the transaction is successful and there is no `BeginSponsoringOp` then there is something  wrong, but if the tx failed, that's fine.

**Comment by 2opremio** [comment]
Alright. Will fix!

**Comment by bartekn** [comment]
Fixed by #2990 and #2993.
